### PR TITLE
Selenium/task3 

### PR DESCRIPTION
**Pull request by rabestro**
# Hurt Me Plenty

For this task, please, use the Selenium WebDriver power, framework unit test, and Page Object concepts. 

## Automate the following script:

- [x] Open https://cloud.google.com/.
- [x] Click the search button at the top of the portal page and enter "Google Cloud Platform Pricing Calculator" into the search field.
- [x] Click on the search button to start searching.
- [x] Click "Google Cloud Platform Pricing Calculator" in the search results and go to the calculator page.
- [x] Click COMPUTE ENGINE at the top of the page.

### Fill out the form with the following data:

- [x] Number of instances: 4
- [x] What are these instances for?: leave blank
- [x] Operating System / Software: Free: Debian, CentOS, CoreOS, Ubuntu, or another User-Provided OS
- [x] VM Class: Regular
- [x] Instance type: n1-standard-8 (vCPUs: 8, RAM: 30 GB)
- [x] Select “Add GPUs”
- [x] Number of GPUs: 1
- [x] GPU type: NVIDIA Tesla V100
- [x] Local SSD: 2x375 Gb
- [x] Datacenter location: Frankfurt (europe-west3)
- [x] Committed usage: 1 Year

### Click Add to Estimate.

Check that the data is correct in the following fields: 

- [x] VM Class, 
- [x] Instance type, 
- [x] Region, 
- [x] local SSD
- [x] the commitment term.

Check that the monthly rent matches the amount received manually when the test is done.

### CODE REVIEW

File: selenium-task-3/src/test/java/com/epam/engx/selenium/task3/SearchAndEstimateTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                .model("computeServer.quantity").set("4")
+                .model("computeServer.os").set("free")
+                .model("computeServer.class").set("regular")
+                .model("computeServer.series").set("n1")
+                .model("computeServer.instance").set("n1-standard-8")
+                .model("computeServer.addGPUs").set("true")
+                .model("computeServer.gpuType").set("NVIDIA Tesla V100")
+                .model("computeServer.gpuCount").set("1")
+                .model("computeServer.ssd").set("2x375")
+                .model("computeServer.location").set("Frankfurt")
```
Comment (yury-barsukou) [inline selenium-task-3/src/test/java/com/epam/engx/selenium/task3/SearchAndEstimateTest.java:67]: Reputable literals like "Frankfurt" is better to store in variables or class constants to reuse and manage easily

**Review by yury-barsukou — APPROVED** [review]

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package com.epam.engx.selenium.pages;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+// page_url = https://cloud.google.com/
```
[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java:8]: @yury-barsukou, this comment has a practical reason: 
https://www.jetbrains.com/help/idea/selenium-page-objects.html#f60bb8aa

> If you want to change the page that opens for a page object, specify the page address in the page_url comment. For FluentLenium, use the [@PageUrl](https://fluentlenium.com/javadoc/org/fluentlenium/core/annotation/PageUrl.html) annotation.

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    @FindBy(name = "q")
+    private WebElement searchInput;
+
+    private GoogleCloud(WebDriver driver) {
+        this.driver = driver;
+        initElements(driver, this);
+    }
+
+    public static GoogleCloud openPage(WebDriver driver) {
+        driver.get("https://cloud.google.com/");
```
[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java:21]: Done

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/SearchResult.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+// page_url = https://cloud.google.com/s/results?q=Google%20Cloud%20Platform%20Pricing%20Calculator
+public class SearchResult {
+    private final WebDriver driver;
+
+    @FindBy(xpath = "//a[b]")
+    private List<WebElement> results;
+
+    @FindBy(xpath = "//a[b][1]")
+    private WebElement firstResult;
```
[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/SearchResult.java:20]: done

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/gcpc/Estimate.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    public Map<String, String> items() {
+        return Stream.of("Region", "Commitment term", "Provisioning model",
+                        "Instance type", "Operating System", "Local SSD")
+                .collect(Collectors.toUnmodifiableMap(Function.identity(), this::value));
+    }
+
+    private String value(String text) {
+        return item(text).lines().findFirst().orElse("")
+                .replaceFirst("[^:]+?: ", "");
+    }
```
[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/gcpc/Estimate.java:43]: Yes, you are right. The names are not informative enough. It is not always easy to choose a suitable name :)

However, I'm not sure that adding a `get` is a good idea. Here is an article I would like to recommend reading: https://www.yegor256.com/2014/09/16/getters-and-setters-are-evil.html

[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/gcpc/Estimate.java:43]: Using `get` is an old tradition. In old java, you can find methods like:
`Striing::getBytes`. However, in new releases, the naming traditions are changed. For example: `String::chars`, `String::codePointAt`, `String::length()`, `String::lines` and so on. 

I think about more expressive names. You point me to the real issue.

File: selenium-task-3/src/test/java/com/epam/engx/selenium/task3/SearchAndEstimateTest.java (hunk 1 above)
Comment (rabestro) [reply selenium-task-3/src/test/java/com/epam/engx/selenium/task3/SearchAndEstimateTest.java:67]: done

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yury-barsukou) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java:8]: Ok I see, but it seems that you don't use this automation step and open the page through the openPage method

File: selenium-pages/src/test/java/com/epam/engx/selenium/pages/gcpc/ByModelTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    @ParameterizedTest(name = "model: {0}")
+    @CsvSource({
+            "listingCtrl.computeServer.os, model(listingCtrl.computeServer.os)",
+            "computeServer.os, model(listingCtrl.computeServer.os)",
+            "os, model(listingCtrl.computeServer.os)",
+            "persistentDisk.location, model(listingCtrl.persistentDisk.location)",
+            "location, model(listingCtrl.computeServer.location)"
+    })
+    void create_a_locator_by_angular_model_name(String model, String locator) {
```
[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/test/java/com/epam/engx/selenium/pages/gcpc/ByModelTest.java:39]: I use a [Display Name Generator](https://www.thewokecoder.io/cool-features-junit5/).

This is what I would discuss with you. I found several projects on GitHub that use this approach. Not too much. Do you think this is a good or bad idea? Should we strictly stick Java method name convention in the tests? Or can we use another convention to improve readability?

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/gcpc/Estimate.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (yury-barsukou) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/gcpc/Estimate.java:43]: I see your point, and it makes sense for sure. But to be well understood in the real world between automation engineers
which are not always too deep in programming as developers and not chasing the latest trends in programming languages (which are a lot in your solutions) you should adjust. Using getters is a classical approach which is not necessarily the best but well familiar and widely used in test automation.

File: selenium-pages/src/test/java/com/epam/engx/selenium/pages/gcpc/ByModelTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (yury-barsukou) [reply selenium-pages/src/test/java/com/epam/engx/selenium/pages/gcpc/ByModelTest.java:39]: we can use another convention if it benefits your project (allow reusing autogenerated methods, for instance), but you should agree on this with your team. And of course, configure your static code analyzers to ignore this kind of violations.

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/GoogleCloud.java]: yes, I created the pages first and only some time I read about this feature and add comment to test it :)

File: selenium-pages/src/main/java/com/epam/engx/selenium/pages/gcpc/Estimate.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rabestro) [reply selenium-pages/src/main/java/com/epam/engx/selenium/pages/gcpc/Estimate.java]: done

**Pull request merged**